src/extract.rs

```diff
@@ -57,75 +59,94 @@ fn extract_links_from_markdown(input: &str) -> Vec<String> {
-                                urls.push(String::from_utf8_lossy(attr.value.as_ref()).to_string());
-                            }
-                            _ => {
-                                for link in extract_links_from_plaintext(
-                                    &String::from_utf8_lossy(attr.value.as_ref()).to_string(),
-                                ) {
-                                    urls.push(link);
-                                }
-                            }
-                        }
-                    }
-                }
+    // we pass mutable urls reference to avoid extra allocations in each
+    // recursive descent
+    walk_html_links(&mut urls, &rc_dom.document);
+
+    urls
+}
+
+/// Recursively walk links in a HTML document, aggregating URL strings in `urls`.
+fn walk_html_links(mut urls: &mut Vec<String>, node: &Handle) {
+    match node.data {
+        NodeData::Text { ref contents } => {
+            // escape_default turns tab characters into "\t", newlines into "\n", etc.
+            let esc_contents = contents.borrow().escape_default().to_string();
+            for link in extract_links_from_plaintext(&esc_contents) {
+                urls.push(link);
             }
-            HTMLEvent::Text(txt) | HTMLEvent::Comment(txt) => {
-                for link in extract_links_from_plaintext(
-                    &String::from_utf8_lossy(txt.escaped()).to_string(),
-                ) {
-                    urls.push(link);
-                }
+        }
+
+        NodeData::Comment { ref contents } => {
+            for link in extract_links_from_plaintext(&contents.escape_default().to_string()) {
+                urls.push(link);
             }
-            HTMLEvent::Eof => {
-                break;
+        }
+
+        NodeData::Element {
+            ref name,
+            ref attrs,
+            ..
+        } => {
+            for attr in attrs.borrow().iter() {
+                let attr_value = attr.value.escape_default().to_string();
+
+                if elem_attr_is_link(attr.name.local.as_ref(), name.local.as_ref()) {
+                    urls.push(attr_value);
+                } else {
+                    for link in extract_links_from_plaintext(&attr_value) {
+                        urls.push(link);
+                    }
+                }
             }
-            _ => {}
         }
-        buf.clear();
+
+        _ => {}
+    }
+
+    // recursively traverse the document's nodes -- this doesn't need any extra
+    // exit conditions because the document is a tree
+    for child in node.children.borrow().iter() {
+        walk_html_links(&mut urls, child);
     }
-    urls
 }
 
-// Extracting unparsed URL strings from a plaintext
+/// Determine if element's attribute contains a link / URL.
+fn elem_attr_is_link(attr_name: &str, elem_name: &str) -> bool {
+    matches!(
+        (attr_name, elem_name),
+        ("href", "a")
```

Comment:
Just a thought: would wildcards work here? E.g. instead of `("href", "a")` we use `("href", _)`. So essentially we assume any `href` is a valid link. Same for `src` and friends. Can't think of any element where this is not the case (and I might be wrong). Or was the idea to be explicit in what we accept? 🤔 
[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=44b796a6d28f4402aa8aa2b59e91adff).

Reply:
Nice idea!

You're right, `href` and `src` seem only to be used for links: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes -- I guess it wouldn't hurt to wildcard these attributes.